cmd/kube-proxy/app/server_test.go

```diff
@@ -235,7 +240,7 @@ nodePortAddresses:
 
 		assert.NoError(t, err, "unexpected error for %s: %v", tc.name, err)
 
-		if !reflect.DeepEqual(expected, config) {
+		if !reflect.DeepEqual(expected, config) && cmp.Diff(config, expected) != "" {
```

Comment:
nit: Why this extra `cmp.Diff`?

This could also be written as:
```
if diff := cmp.Diff(config, expected); diff != "" {
    t.Fatalf("unexpected config for %s, diff = %s", tc.name, diff)
}
```

Reply:
sometimes `DeepEqual` is false but `Diff` is empty, which causes the unit test to fail. Yes, Using `Cmp.Diff` is perhaps simpler and more accurate.

Reply:
Can you give an example? Is such a corner case a failure or something that `Diff` doesn't handle (properly)?

Reply:
```
git diff
...
-               if cmp.Diff(config, expected) != "" {
+               if !reflect.DeepEqual(expected, config) {
                        t.Fatalf("unexpected config for %s, diff = %s", tc.name, cmp.Diff(config, expected))
                }
        }
go test -v  ./cmd/kube-proxy/app/ -run TestLoadConfig 
=== RUN   TestLoadConfig
    server_test.go:244: unexpected config for iptables mode, IPv4 all-zeros bind address, diff = 
--- FAIL: TestLoadConfig (0.00s)
=== RUN   TestLoadConfigFailures
=== RUN   TestLoadConfigFailures/Decode_error_test
=== RUN   TestLoadConfigFailures/Bad_config_type_test
=== RUN   TestLoadConfigFailures/Missing_quotes_around_::_bindAddress
--- PASS: TestLoadConfigFailures (0.00s)
    --- PASS: TestLoadConfigFailures/Decode_error_test (0.00s)
    --- PASS: TestLoadConfigFailures/Bad_config_type_test (0.00s)
    --- PASS: TestLoadConfigFailures/Missing_quotes_around_::_bindAddress (0.00s)
FAIL
FAIL    k8s.io/kubernetes/cmd/kube-proxy/app    0.950s
FAIL

```

Reply:
Why does `reflect.DeepEqual` return false here? Is that a bug in `DeepEqual`? If yes, it should be fixed. If no, then the test should continue to use `DeepEqual` and we need to investigate why `cmp.Diff` doesn't report the difference.

Reply:
yes, I'm investigating what's going on here

Reply:
I found the reason, some default values are missing in `expected.Logging`,  see https://github.com/kubernetes/kubernetes/compare/3c35b0dddfb28538686b8961e23a98e954f6070e..18a5dd572463f4df7e62196f47f5c5cc7cd691e7

Reply:
So `DeepEqual` failed because `expected.Logging.Options.JSON.InfoBufferSize.String` mutates the quantity and that led to a difference that was caught only be `DeepEqual`? Why not by `cmp.Diff`? Ah, probably because `cmp.Diff` supports calling `Quantity.Equal`.

I think using only `cmp.Diff` is the right solution here. It clearly does a better job than `DeepEqual` and the code becomes simpler:
```
if diff := cmp.Diff(config, expected); diff != "" {
    t.Fatalf("unexpected config for %s, diff = %s", tc.name, diff)
}
```

Let's not call it twice, BTW.

With that change, you can revert https://github.com/kubernetes/kubernetes/compare/3c35b0dddfb28538686b8961e23a98e954f6070e..18a5dd572463f4df7e62196f47f5c5cc7cd691e7 because we don't have to work around the `DeepEqual` limitation.

Reply:
`cmp.Diff` is supports calling `Quantity.Equal`? 🤔

OK, let's revert these changes...

Reply:
I supports calling *any* Equal method, not `Quantity.Equal` specifically.